### ISSUE TITLE
Bug: QueryBuilder inconsistent behaviour with multiple `set()`s

### ISSUE THREAD

**Issue by tangix** [codeigniter4/CodeIgniter4#7143]
### PHP Version

8.1

### CodeIgniter4 Version

4.3.1

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

macOS

### Which server did you use?

apache

### Database

MySQL 8

### What happened?

Using multiple `set()`s with `RawSql` doesn't produce consistent sql statements.
```
Working SQL (1):
INSERT INTO `auth_bearer` (`jti`, `proctorID`, `token_type`, `expires`) VALUES ('jti', 12, 'handover', DATE_ADD(NOW(), INTERVAL 2 HOUR))
Working SQL (2):
INSERT INTO `auth_bearer` (`jti`, `proctorID`, `token_type`, `expires`) VALUES ('jti', 12, 'handover', DATE_ADD(NOW(), INTERVAL 2 HOUR))
Working SQL (3):
INSERT INTO `auth_bearer` (`jti`, `proctorID`, `token_type`, `expires`) VALUES ('jti', 12, 'handover', DATE_ADD(NOW(), INTERVAL 2 HOUR))
Generates bad SQL:
INSERT INTO `auth_bearer` (`jti`, `proctorID`, `token_type`, expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)) VALUES ('jti', 12, 'handover', )
Silently failing SQL:
INSERT INTO `auth_bearer` (`jti`, `proctorID`, `token_type`) VALUES ('jti', '12', 'handover')
Failing with exception: You must use the "set" method to insert an entry.
```

### Steps to Reproduce

The following code produce different sql statements, some working and some not working:

```php
       $db = db_connect();

        CLI::write('Working SQL (1):', 'green');
        CLI::write($db->table('auth_bearer')
        ->set([
            'jti'        => 'jti',
            'proctorID'  => 12,
            'token_type' => 'handover',
        ])        
        ->set('expires', 'DATE_ADD(NOW(), INTERVAL 2 HOUR)', false)
        ->getCompiledInsert(true));

        CLI::write('Working SQL (2):', 'green');
        CLI::write($db->table('auth_bearer')
        ->set([
            'jti'        => 'jti',
            'proctorID'  => 12,
            'token_type' => 'handover',
            'expires' => new RawSql('DATE_ADD(NOW(), INTERVAL 2 HOUR)')
        ])        
        ->getCompiledInsert(true));

        CLI::write('Working SQL (3):', 'green');
        CLI::write($db->table('auth_bearer')
        ->set('jti', 'jti')
        ->set('proctorID', 12)
        ->set('token_type', 'handover')
        ->set('expires', 'DATE_ADD(NOW(), INTERVAL 2 HOUR)', false)
        ->getCompiledInsert(true));

        CLI::write('Generates bad SQL:', 'red');
        CLI::write($db->table('auth_bearer')
        ->set([
            'jti'        => 'jti',
            'proctorID'  => 12,
            'token_type' => 'handover',
        ])        
        ->set('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)', null, false)
        ->getCompiledInsert(true));


        CLI::write('Silently failing SQL:', 'red');
        CLI::write($db->table('auth_bearer')
        ->set([
            'jti'        => 'jti',
            'proctorID'  => '12',
            'token_type' => 'handover',
        ])        
        ->set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
        ->getCompiledInsert(true));


        try {
            CLI::write($db->table('auth_bearer')
            ->set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
            ->getCompiledInsert(true));
        }
        catch (DatabaseException $e) {
            CLI::write('Failing with exception: ' . $e->getMessage(), 'red');
        }
```

### Expected Output

Same statements in all valid cases.

**Comment by kenjis** [comment]
> This method enables you to set values for inserts or updates.
https://codeigniter4.github.io/CodeIgniter4/database/query_builder.html#builder-set

I understand `set()` sets key value pairs.

The following does not set key value pairs.
```php
set('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)', null, false)
set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
```

**Comment by tangix** [comment]
So maybe that's why the call discussed in #7117 works (because it uses a blank string for `$value`)
```
set('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)', '', false)
```

**Comment by kenjis** [comment]
> Is `$QB->set($data)->insert()` not supported but should be `$QB->insert($data)` instead? 

`$QB->set($data)->insert()` is supported.
It is clearly documented: https://codeigniter4.github.io/CodeIgniter4/database/query_builder.html#builder-set

**Comment by tangix** [comment]
OK, I'll continue to use `(3)` in my code.

**Comment by kenjis** [comment]
The following code also should work:
```php
$db->table('auth_bearer')
        ->set('jti', 'jti')
        ->set('proctorID', 12)
        ->set('token_type', 'handover')
        ->set('expires', new RawSql('DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
```

**Comment by kenjis** [comment]
@codeigniter4/database-team 

The following code
```php
        $builder = $this->db->table('auth_bearer');

        $sql = $builder
            ->set([
                'jti'        => 'jti',
                'proctorID'  => '12',
                'token_type' => 'handover',
            ])
            ->set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
            ->getCompiledInsert();
```
generates without any error:
```sql
INSERT INTO "auth_bearer" ("jti", "proctorID", "token_type") VALUES ('jti', '12', 'handover')
```
This is not good.

**Comment by tangix** [comment]
> This is not good.

One could argue that that code is bad, but the docs https://codeigniter.com/user_guide/database/query_builder.html#CodeIgniter\Database\BaseBuilder::set are not clear on that.
Maybe this is all due to not-perfect docs and me relying on bad coding practices?

**Comment by MGatner** [comment]
Is the problem the multiple `set()` statements? Or that the Raw SQL doesn't include a value? Seems like either way we should be getting an exception here instead of disappearing SQL input.

**Comment by tangix** [comment]
@MGatner regarding the Exception: since this code throws an `CodeIgniter\Database\Exceptions\DatabaseException`:
```
        CLI::write('Throws Exception', 'green');
        CLI::write($db->table('auth_bearer')
        ->set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)')) // not working
        ->getCompiledInsert(true));
```
So should this code:
```
        CLI::write('Should throw Exception', 'red');
        CLI::write($db->table('auth_bearer')
        ->set('jti', 'jti')
        ->set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)')) // not working
        ->getCompiledInsert(true));
```

**Comment by tangix** [comment]
@MGatner as to multiple `set()`, I am fine with either way here. The `RawSql` helped be with problematic `BETWEEN` in #7117 so I thought that should help me here as well... My understanding of how the framework reacts to `RawSql` seems to have some limits.
Thinking more and checking my code I think the issue I'm having boils down to how `set()` and `where()` differs:
```php
// Not working!
CLI::write($db->table('auth_bearer')
        ->select('*')
        ->where('expires', new RawSql('DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
        ->getCompiledSelect(true));

// Working!
CLI::write($db->table('auth_bearer')
        ->select('*')
        ->where(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
        ->getCompiledSelect(true));

```
These two statements produce:
```sql
SELECT *
FROM `auth_bearer`
WHERE DATE_ADD(NOW(), INTERVAL 2 HOUR)

SELECT *
FROM `auth_bearer`
WHERE expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)
```
which is inconsistent with `set()`:

```php
// Working!
CLI::write($db->table('auth_bearer')
        ->set('expires', new RawSql('DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
        ->getCompiledInsert(true));

// Not working - throwing exception
CLI::write($db->table('auth_bearer')
        ->set(new RawSql('expires = DATE_ADD(NOW(), INTERVAL 2 HOUR)'))
        ->getCompiledInsert(true));
```
`¯\_(ツ)_/¯`

**Comment by sclubricants** [comment]
@tangix I submitted a PR to fix the functionality of `where()` so that it behaves the same way `set()` does. 

You can set both key and value to use RawSql.

```php
        $sql = $this->db->table('auth_bearer')
            ->select('*')
            ->where(new RawSql('CURRENT_TIMESTAMP() >='), new RawSql('DATE_ADD(column, INTERVAL 2 HOUR)'))
            ->getCompiledSelect(true);

        $expected = <<<'SQL'
            SELECT *
            FROM "auth_bearer"
            WHERE CURRENT_TIMESTAMP() >= DATE_ADD(column, INTERVAL 2 HOUR)
            SQL;
        $this->assertSame($expected, $sql);
```
You can include an operator in the key or leave it out for =

You can also use only RawSql for key and leave value null.

**Comment by kenjis** [comment]
I sent a PR to fix Silently failing SQL https://github.com/codeigniter4/CodeIgniter4/issues/7143#issuecomment-1385258336
See #7148

**Comment by tangix** [comment]
@sclubricants 
Thanks! As I understand it my bad code worked until now so be careful not to break anything with that PR... Compared to framework devs, we users are very sensitive to that...

My bad code must have been introduced in 4.0 (this project is not ported from 3.x to 4.x, it was created from scratch). Maybe the docs need some ❤️ as well, I'll have a look at that.

**Comment by kenjis** [comment]
Closed by #7147 and #7148
